### ISSUE TITLE
Type equalities 

### ISSUE THREAD

**Issue by nikivazou** [ucsd-progsys/liquidhaskell#701]
Type equalities are treated as class constraints (as of recently). So the [following](https://github.com/ucsd-progsys/liquidhaskell/blob/ackermann/tests/pos/TypeEqualities.hs) test passes

```
{-# LANGUAGE TypeFamilies   #-}

{-@ zoo :: x:a -> y:a -> {v:b | v ~~ x} @-}
zoo :: (a~b) => a -> a -> b 
zoo x _ = x 
```

Next step is to push the equality information in the logic and allow replacing `v~~x` with `v == x`. The second is not malformed.

**Comment by gridaphobe** [comment]
Is there a reason to push the type equality into the logic? (Besides the ugliness of `~~`?)

**Comment by nikivazou** [comment]
Yes! I want to also have `<` etc, not only `~~`

**Comment by ranjitjhala** [comment]
Can you think of a way to NOT push type equality into the logic? I don't
think that's a good idea.

**Comment by nikivazou** [comment]
I think we need to reason about all the type class info properly: right now we have a hack for numerics, we should generalize it with a better story

**Comment by ranjitjhala** [comment]
Not seeing the connection of type classes  and equality constraints ...

**Comment by nikivazou** [comment]
They are very similar
   - arguments are inferred by the compiler
   - they are printed before `=>` in the type signature 
   - both are represented as type constructors applied to types 


Actually, I expected `~` to be a class constraint, but it is not

**Comment by gridaphobe** [comment]
Yes, class constraints are very different from equality constraints. (`<` would count as a class constraint)

**Comment by gridaphobe** [comment]
Specifically, the fancier constraints, like the arithmetic operators could be useful to push into the logic, but the equality constraints seem pointless to me.

**Comment by nikivazou** [comment]
@gridaphobe I want to be able to write 

```
{-# LANGUAGE TypeFamilies   #-}

{-@ zoo :: x:a -> y:a -> {v:b | v < x} @-}
zoo :: (a~b) => a -> a -> b 
zoo x _ = x
```

**Comment by ranjitjhala** [comment]
@nikivazou, *why* do you want to write that? Whats wrong with:

```

{-@ zoo :: x:a -> y:a -> {v:b | v < x} @-}
zoo :: (a~b) => a -> a -> b
zoo x _ = x

```


or at any rate, for LH emitting FQ constraints that would be

equivalent to what we get from the above?

**Comment by nikivazou** [comment]
@ranjitjhala under an environment that does not know `a~b`, the refinement you wrote is unsorted, as it reduces to  `x:a, v:b |- v < x`

**Comment by gridaphobe** [comment]
1. Why do you want to write that? In my experience writing an explicit equality constraint in the context is a bad sign.

2. Why not just apply the equality before running the sort-checker?

**Comment by nikivazou** [comment]
I did not understand your 2 point. 

The reason I want to do that is here 
https://github.com/ucsd-progsys/liquidhaskell/pull/700/commits/44637d5eb2b961979012e9c0c78b3bb4aab07a7b

but I can see more Haskell users asking this feature.

**Comment by gridaphobe** [comment]
You can treat the equality as a substitution `a := b`, to be applied before sort-checking. Then there's no need to deal with equality constraints in the logic.

**Comment by nikivazou** [comment]
That is a nice solution!

**Comment by ranjitjhala** [comment]
@nikivazou -- sorry I cut paste without editing, I meant just replace all
`a` with `b` or vice versa...

**Comment by ranjitjhala** [comment]
but to do that in *LH* not inside fixpoint.

**Comment by nikivazou** [comment]
So, more or less what Eric says. It is a solution!
